Fix: `is-equal` fails in some cases when intersecting wrapped types `(F<[A]> & G<[A]>)`

taiyakihitotsu opened the pull request:
<!--

Thanks for submitting a pull request 🙌

If you're submitting a new type, please review the contribution guidelines:
https://github.com/sindresorhus/type-fest/blob/main/.github/contributing.md

-->

Hello.

This PR fixes an edge case of `is-equal.d.ts`:
(Sorry if I've made any mistakes with the PR or coding, and please forgive me if my English isn't very good...)

Assuming `F` and `G` return `['something', true]` if the `extends` condition is matched, or `['something', false]` if not.
If we want to intersect thir return types `(F<A> & G<A>)` and expect to get `never` so that `IsEqual<It, never>` evaluates to `true`, but it instead returns `false` when `A` is wraped as `[A]` and the right side of the `extends` in both `F` and `G` is also wrapped.

This is the case I want to fix.

This case occurs, as shown below, when a type variable is introduced via infer.

A concrete example is easier to understand, so I will provide an overview below.

```typescript
// // [false, never] case
type __TupleRetTupleBool_wraped0<Fst> = (Fst extends [[0, 2]] ? ['A', true] : ['A', false])
type __TupleRetTupleBool_wraped1<Fst> = (Fst extends [[0, 1]] ? ['A', true] : ['A', false])

type FFF_wraped<Fst> = [IsEqual<(__TupleRetTupleBool_wraped0<[Fst]> & __TupleRetTupleBool_wraped1<[Fst]>), never>, (__TupleRetTupleBool_wraped0<[Fst]> & __TupleRetTupleBool_wraped1<[Fst]>)]
type __TupleRetTupleBool_wrapedr = __TupleRetTupleBool_wraped0<[[0,2]]> // ['A', true]
type __TupleRetTupleBool_wraped1r = __TupleRetTupleBool_wraped1<[[0,2]]> // ['A', false]

type check_FFF_wraped = FFF_wraped<[0,2]> // [false,never] before fix.
type expand_FFF_wraped = IsEqual<(__TupleRetTupleBool_wraped0<[[0,2]]> & __TupleRetTupleBool_wraped1<[[0,2]]>), never> // true before/after fix.
type infer_FFF_wraped = [0,2] extends infer Fst ? IsEqual<(__TupleRetTupleBool_wraped0<[Fst]> & __TupleRetTupleBool_wraped1<[Fst]>), never> : never // false before fix.

// // [true, never] case
type __TupleRetTupleBool0<Fst> = (Fst extends [0, 2] ? ['A', true] : ['A', false])
type __TupleRetTupleBool1<Fst> = (Fst extends [0, 1] ? ['A', true] : ['A', false])

type FFF_notwrap<Fst> = [IsEqual<(__TupleRetTupleBool0<Fst> & __TupleRetTupleBool1<Fst>), never>, (__TupleRetTupleBool0<Fst> & __TupleRetTupleBool1<Fst>)]
type __TupleRetTupleBool0r = __TupleRetTupleBool0<[0,2]> // ['A', true]
type __TupleRetTupleBool1r = __TupleRetTupleBool1<[0,2]> // ['A', false]

type check_FFF_notwrap = FFF_notwrap<[0,2]> // [true,never] before/after fix.
type expand_FFF = IsEqual<(__TupleRetTupleBool0<[0,2]> & __TupleRetTupleBool1<[0,2]>), never> // true before/after fix.
type infer_FFF = [0,2] extends infer Fst ? IsEqual<(__TupleRetTupleBool0<Fst> & __TupleRetTupleBool1<Fst>), never> : never // true before/after fix.
```

Thanks in advance!

som-sm commented:
@taiyakihitotsu Even the following solution seems to pass all the existing test cases:

```ts
export type IsEqual<A, B> =
	[A] extends [infer A]
		? [B] extends [infer B]
			? _IsEqual<A, B>
			: false
		: false;

type _IsEqual<A, B> =
	(<G>() => G extends A & G | G ? 1 : 2) extends
	(<G>() => G extends B & G | G ? 1 : 2)
		? true
		: false;
```

So, either we can simplify the existing solution or we need to add tests that fail with the above solution.

som-sm commented:
> If I decide to adopt your code, should I just leave a comment to mention that? Or is there something I should do on Git to reflect that properly? Or else?
> 
> Let me know if there's a preferred way to reflect your suggestion properly — I just want to make sure it doesn’t look like I’m claiming credit for it.

No attribution required, feel free to just copy paste the code. 

If you really want to, you can add [co-authors in your commits](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors). But it's also fine the way you have done it.

som-sm commented:
> If I decide to adopt your code, should I just leave a comment to mention that? Or is there something I should do on Git to reflect that properly? Or else?
> 
> Let me know if there's a preferred way to reflect your suggestion properly — I just want to make sure it doesn’t look like I’m claiming credit for it.

No attribution required, feel free to just copy paste the code. 

If you really want to, you can add [co-authors in your commits](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors).

taiyakihitotsu commented:
@som-sm 

Understood.
I will add the co-authors when I work on it later.
Please let me know if there are any issues afterwards.

som-sm commented:
> Understood.
> I will add the co-authors when I work on it later.
> Please let me know if there are any issues afterwards.

If you're planning to me as a co-author, then you'd need my email. Use:

```
 Co-authored-by: Som Shekhar Mukherjee <49264891+som-sm@users.noreply.github.com>
 ```

taiyakihitotsu commented:
I found and added those three test cases, along with one simple baseline case (`[never], [never]`).
(`IsEqual` is by @som-sm, and `IIIsEqual` is my own implementation, which I commited.)

These tests demonstrate that @som-sm's definition is correct.

```typescript
type Greet = {
  (fullName: string): string;
  (firstName: string, lastName: string): string;
};

type InferGreet = Greet extends infer F ? F extends (0 extends 0 ? F : 1) ? [F] : 2 : 3

type IIexpectTrueIntersectionGreet = IIIsEqual<[Greet] & [Greet], InferGreet>// false
type expectTrueIntersectionGreet = IsEqual<[Greet] & [Greet], InferGreet> // true

type StringArrayReturn = {
  (s: string): string[];
  (ss: string[]): string[];
};
type InferStringArrayReturn = StringArrayReturn extends infer F ? F extends (0 extends 0 ? F : 1) ? [F] : 2 : 3

type IIexpectTrueIntersectionStringArrayReturn = IIIsEqual<[StringArrayReturn] & [StringArrayReturn], InferStringArrayReturn> // false
type expectTrueIntersectionStringArrayReturn = IsEqual<[StringArrayReturn] & [StringArrayReturn], InferStringArrayReturn> // true

type SomeTypeReturn = {
  <T>(value: T): T;
  <T>(value: T[]): T[];
};
type InferSomeTypeReturn = SomeTypeReturn extends infer F ? F extends (0 extends 0 ? F : 1) ? [F] : 2 : 3

type IIexpectTrueIntersectionSomeTypeReturn = IIIsEqual<[SomeTypeReturn] & [SomeTypeReturn], InferSomeTypeReturn> // false
type expectTrueIntersectionSomeTypeReturn = IsEqual<[SomeTypeReturn] & [SomeTypeReturn], InferSomeTypeReturn> // true
```

(This may be due to the fact that intersection types like `[T] & [T]` are not simplified in TypeScript.)

I also added @som-sm as a co-author. Please take a look.

som-sm commented:
> I'm a bit confused, all these three new tests seem to pass with all the implementations of `IsEqual`, i.e. these tests don't fail with the current implementation, nor with the ones you and I suggested.

@taiyakihitotsu I'm still not sure if those test cases are required. If not, let's remove them. If they are required, can you tell me with which implementation of `IsEqual` those cases fail?

Everything else LGTM!

taiyakihitotsu commented:
> In that case, I don't think this is related to function overloading, looks like the test could be simplified to this:

Thanks, I didn’t know that.

I had assumed that both narrowing and intersection were involved in determining the test result, since in the three test cases above, the outcome varied depending on whether `declare const` or `const` was used — as was pointed out in a previous review.

If the test can be sufficiently covered using just intersection, I think that should be fine too.

taiyakihitotsu commented:
@som-sm

I pushed this change. https://github.com/sindresorhus/type-fest/pull/1231#issuecomment-3313583257
